keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,337 @@
+
+## Motivation
+
+Different applications may have different business values, different logic and may require different scaling behaviours.
+I can name at least three types of applications:
+
+- Applications that handles business-critical web traffic. They should scale up as fast as possible (false positive signals to scale up are ok), and scale down very slowly (waiting for another traffic spike).
+- Applications that process very important data. They should scale up as fast as possible (to reduce the data processing time), and scale down as soon as possible (to reduce the costs). False positives signals to scale up/down are ok.
+- Applications that process other data/web traffic. It is not that important, and may scale up and down in a usual way to minimize jitter.
+
+At the moment, there’s only one cluster-level configuration parameter that influence how fast the cluster is scaled down:
+
+- [--horizontal-pod-autoscaler-downscale-stabilization-window][]   (default to 5 min)
+
+And a couple of hard-coded constants that specify how fast the cluster can scale up:
+
+- [scaleUpLimitFactor][] = 2.0
+- [scaleUpLimitMinimum][] = 4.0
+
+As a result users can't influence the scale velocity and that is a problem for a lot of people. There're several open issues in tracker about that:
+
+- [#39090][]
+- [#65097][]
+- [#69428][]
+
+[--horizontal-pod-autoscaler-downscale-stabilization-window]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+[scaleUpLimitFactor]: https://github.com/kubernetes/kubernetes/blob/v1.13.4/pkg/controller/podautoscaler/horizontal.go#L55
+[scaleUpLimitMinimum]: https://github.com/kubernetes/kubernetes/blob/v1.13.4/pkg/controller/podautoscaler/horizontal.go#L56
+[#39090]: https://github.com/kubernetes/kubernetes/issues/39090
+[#65097]: https://github.com/kubernetes/kubernetes/issues/65097
+[#69428]: https://github.com/kubernetes/kubernetes/issues/69428
+
+### Goals
+
+- Allow the user to be able to manage the scale velocity
+
+### Non-Goals
+
+TBA
+
+## Proposal
+
+We need to introduce an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behaviour.
+For both direction (scale up and scale down) there should be two parameters:
+
+- Parameter to specify the relative speed, in percentages:
+  - `ScaleUpPercent`
+    i.e. if ScaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
+  - `ScaleDownPercent`
+    i.e. if ScaleDownPercent = 60 , then we can remove 60% of pods (10 -> 4)
+- Parameter to specify the absolute speed, in number of pods:
+  - `ScaleUpPods`
+    i.e. if ScaleUpPods = 5 , then we can add 5 more pods (10 -> 15)
+  - `ScaleDownPods`
+    i.e. if ScaleDownPods = 7 , then we can remove 7 pods (10 -> 3)
+
+All the parameters are per-minute and allow fraction values by using Quantity type.
+
+A user will specify the parameters for the HPA, thus controlling the HPA logic.
```

Comment:
On the PR I asked why not int32, the answer was values 0 < x < 1.

I asked why not specify the denominator? 

We should come up with a convention for rational values.  I think I'd rather see something like:

```
{
    ScaleUp HPAConstraintValue
    ScaleDown HPAConstraintValue
}

type HPAConstraintValue struct {
  // One of the following must be set, but for APi compat reasons it's possible that none may be set, which is interpreted as <somethign, you tell me...>
  Pods int32
  Percent int32

  PeriodSeconds int32
}
```

so you can say:

```
constraints:
  - scaleUp:
      pods: 3
      periodSeconds: 1
  - scaleDown:
      pods: 1
      periodSeconds: 3
```

This may or may not be done anywhere else yet, though I know we discussed somethign similar for CPVPA (@justinsb)

We should consider precedent carefully, but Quantity seems wrong here.

@liggitt for discussion

Reply:
prior art for expressing absolute values or percentages:
- PodDisruptionBudget (`minAvailable: 1` or `minAvailable: "x%"`)
- Deployment rollout strategy (`maxUnavailable: 1` or `maxUnavailable: "x%"`)

Reply:
We should be able to specify both options (absolute and relative) as it may help to configure small clusters as well as large clusters.
I've reworked the KEP according to your suggestion.
Now I'm adding changes to the PR itself https://github.com/kubernetes/kubernetes/pull/74525

Reply:
done, could you check it plz?